cmd/starter/c/starter.c

```diff
@@ -1318,10 +1320,47 @@ __attribute__((constructor)) static void init(void) {
 
         /* engine requested to propagate mount to container */
         if ( sconfig->starter.masterPropagateMount && userns != ENTER_NAMESPACE ) {
+            struct stat rootfs, newrootfs;
+
+            /* keep stat information for root filesystem comparison */
+            if ( stat("/", &rootfs) < 0 ) {
+                fatalf("Failed to get root directory information: %s", strerror(errno));
+            }
+
             /* join child shared mount namespace with relative path */
             if ( enter_namespace("ns/mnt", CLONE_NEWNS) < 0 ) {
                 fatalf("Failed to enter in shared mount namespace: %s\n", strerror(errno));
             }
+
+            /* take stat information after namespace join */
+            if ( stat("/", &newrootfs) < 0 ) {
+                fatalf("Failed to get root directory information: %s", strerror(errno));
+            }
+
+            /*
+             * we compare st_dev and st_ino to check if we are in the current root
+             * filesystem, on some systems the mount namespace join above could escape the
+             * current root filesystem when an init process (initrd, container or chrooted process)
+             * do a chroot instead of switch_root for the initrd case or didn't use
+             * pivot_root/mount MS_MOVE for the container solution case
+             */
+            if ( rootfs.st_dev != newrootfs.st_dev || rootfs.st_ino != newrootfs.st_ino ) {
+                struct statfs fs;
+
+                debugf("Root filesystem change detected, retrieving new root filesystem information\n");
+                if ( statfs("/", &fs) < 0 ) {
+                    fatalf("Failed to retrieve root filesystem information: %s", strerror(errno));
+                }
+
+                /* check if we are in the ram disk filesystem */
+                if ( newrootfs.st_ino == 2 && (fs.f_type == RAMFS_MAGIC || fs.f_type == TMPFS_MAGIC) ) {
```

Comment:
Do we really need to get so specific as to check for the initrd here? This is unlikely to be well tested, and we could just say "Running inside a weak chrooted environment, prefer pivot_root / switch_root instead of chroot" instead of complicating our checks?

Reply:
Discussed - can keep this.